finality-aleph/src/data_io.rs

```diff
@@ -203,138 +448,535 @@ where
+    }
+
+    fn on_message_dependencies_resolved(&self, message: Message) {
+        trace!(target: "afa", "Sending message from DataStore {:?}", message);
+        if let Err(e) = self.messages_for_aleph.unbounded_send(message) {
+            error!(target: "afa", "Unable to send a ready message from DataStore {}", e);
+        }
+    }
+
+    fn assign_fresh_message_id(&mut self) -> MessageId {
+        self.next_free_id += 1;
+        self.next_free_id - 1
+    }
+
+    // Add a message to pending messages. The set of pending proposals is guaranteed to be
+    // nonempty, and the proposals in the set have already been added, but they were not yet
+    // linked with the message.
+    fn add_pending_message(
+        &mut self,
+        message: Message,
+        pending_proposals: HashSet<AlephProposal<B>>,
+    ) {
+        let id = self.assign_fresh_message_id();
+        for proposal in &pending_proposals {
+            let proposal_status = self
+                .pending_proposals
+                .get_mut(proposal)
+                .expect("pending data must be available");
+            proposal_status.messages.insert(id);
+        }
+        let message_status = PendingMessageInfo::new(message, pending_proposals);
+        self.pending_messages.insert(id, message_status);
+    }
 
-        if requirements.is_empty() {
-            trace!(target: "afa", "Sending message from DataStore {:?}", message);
-            if let Err(e) = self.messages_for_aleph.unbounded_send(message) {
-                debug!(target: "afa", "Unable to send a ready message from DataStore {}", e);
+    // This is called upon a proposal being available -- we remove it from the set of
+    // proposals a message waits for.
+    fn remove_proposal_from_pending_message(&mut self, proposal: &AlephProposal<B>, id: MessageId) {
+        let message_info = self
+            .pending_messages
+            .get_mut(&id)
+            .expect("message must exist if a proposal refers to it");
+        message_info.pending_proposals.remove(proposal);
+        if message_info.pending_proposals.is_empty() {
+            let message = self
+                .pending_messages
+                .remove(&id)
+                .expect("message must exist if a proposal refers to it")
+                .message;
+            self.on_message_dependencies_resolved(message);
+        }
+    }
+
+    fn remove_id_from_pending_proposal(&mut self, proposal: &AlephProposal<B>, id: MessageId) {
+        use std::collections::hash_map::Entry::Occupied;
+
+        if let Occupied(mut proposal_entry) = self.pending_proposals.entry(proposal.clone()) {
+            let proposal_info = proposal_entry.get_mut();
+            proposal_info.messages.remove(&id);
+            if proposal_info.messages.is_empty() {
+                proposal_entry.remove();
+            }
+        } else {
+            error!(target:"afa", "Proposal {:?} with id {:?} referenced in message does not exist", proposal, id);
```

Comment:
This implies a bug in the code, still should be a `warn` only?

Reply:
Iirc `error` was supposed to be reserved for situations when the component in question cannot continue operation (not necessarily a `panic`, because it might e.g. just be within one session. `warn` is for when something that should not happen during normal execution happened, but we can continue nonetheless.

Reply:
Fair. Changed to `warn`